pallets/proof_of_existence/src/lib.rs

```diff
@@ -121,6 +138,18 @@ pub mod pallet {
 
             Self::deposit_event(Event::NewAttestation { id, attestation });
 
+            // Prune old attestations
+            // Rationale: ids are incremental, no more than one attestation
+            // for each publish_attestation call will need to be removed from storage
+            let max_attestations = T::MaxStorageAttestations::get();
+            if max_attestations != MaxStorageAttestations::default()
+                && id + 1 >= max_attestations.into()
```

Comment:
Maybe I'm missing something here, but it looks like this would likely break if we change (either way) the value of `MaxStorageAttestations` with a runtime update?
I'd say that we need to have a loop which also takes into account `oldest_attestation_id` to have a robust implementation?

Reply:
Yes. We iterated a bit on this and this is the most minimal solution that wouldn't make the weights increase too much. But I agree that it won't work if, for instance, we decrease the MaxAttestations parameter. Let me see if I can find a way to fix this in a cheaper way, but keep in mind that we will be refactoring this pallet heavily in the upcoming sprint

Reply:
Just a little comment about that: we can meet this issue in just when on update runtime when the attestations in the storage are (much) bigger than  `MaxStorageAttestations`.

Now in the next upgrade we'll use a number that will be bigger than the attestation already present in the storage.

Can a more robust implementation save us later when we'll change this number drastically in a future by a runtime update?

I think the answer is NO: if we implement we cannot benchmarking the worst case but just the normal one. In this case when we issue the extrinsic and we need to remove something like 50000 attestation we have a lot of chances to kill the chain because this transaction will take all block chain to run but the estimated one is just for remove one attestation.

What I really mean here is that we cannot implement a variable behavior in our extrinsic and if we really need to do it we need to estimate the gas at least in all worst condition and in this case really we cannot do it.

What we can do is to write a migration for the pallet that will be run on runtime upgrade and without weight constrains. We should just take care of it when we'll perform an upgrade that reduce this value: my advice is to add a comment or/and (better)  add a static assert that should be greater-equal to that value to raise the red flag in the case we reduce this number.

Anyway I guess that we'll change this pallet before we'll run in this issue.

Reply:
I've added more robust logic and a test case, just to close this.
Most likely we will be refactoring it in the future anyway
https://github.com/HorizenLabs/zkVerify/pull/124/commits/56cdb79994fe4bc825465bf6e4bfb848470af7bf

Reply:
Thank you for the update. I totally agree with the comment that you added to the code, and that we should have some limits on the actual number of iterations that can be performed in reality to avoid timeout issues. Or maybe, we could consider refactoring that "cleanup" functionality into some dedicated extrinsic, to also account more predictably for the fee incurred by the extrinsic.

Anyways, given that we are already planning more changes and refactoring, I'm ok with what we have, for now.